sotodlib/preprocess/preprocess_util.py

```diff
@@ -482,17 +488,37 @@ def multilayer_load_and_preprocess(obs_id, configs_init, configs_proc,
             if 'valid_data' in aman.preprocess:
                 aman.preprocess.move('valid_data', None)
             aman.preprocess.merge(proc_aman.preprocess)
-            pipe_proc.run(aman, aman.preprocess, select=False)
+            if stop_for_sims:
+                batch_idx = [
+                    (step, process.name)
+                    for step, process in enumerate(pipe_proc)
+                    if process.use_data_aman
+                ]
+                batch_idx = [(0, pipe_proc[0].name)] + batch_idx
+                pipes = {}
+                for idx in range(len(batch_idx)-1):
+                    start, start_name = batch_idx[idx]
+                    end, end_name = batch_idx[idx+1]
+                    pipes[end, end_name] = pipe_proc[start:end]
+
+                out_amans = {}
+                loc_aman = aman.copy()
+                for (step, name), pipe in pipes.items():
+                    pipe.run(loc_aman, aman.preprocess, select=False)
+                    out_amans[step, name] = loc_aman.copy()
+                return out_amans
```

Comment:
Could this just be a dictionary of numpy arrays or restricted AxisManagers that only include the fields you need to reduce memory overhead?